kms/mackms/mackms.go

```diff
@@ -316,6 +322,150 @@ func (k *MacKMS) CreateSigner(req *apiv1.CreateSignerRequest) (crypto.Signer, er
+// Valid names (URIs) are:
+//   - mackms:label=test@example.com
+//   - mackms:serial=2c273934eda8454d2595a94497e2395a
+//   - mackms:label=test@example.com;serial=2c273934eda8454d2595a94497e2395a
+func (k *MacKMS) LoadCertificateChain(req *apiv1.LoadCertificateChainRequest) ([]*x509.Certificate, error) {
+	if req.Name == "" {
+		return nil, fmt.Errorf("loadCertificateChainRequest 'name' cannot be empty")
+	}
+
+	// Require label or serial
+	u, err := parseCertURI(req.Name, true)
+	if err != nil {
+		return nil, fmt.Errorf("mackms LoadCertificateChain failed: %w", err)
+	}
+
+	cert, err := loadCertificate(u.label, u.serialNumber, nil)
+	if err != nil {
+		return nil, fmt.Errorf("mackms LoadCertificateChain failed1: %w", err)
+	}
+
+	chain := []*x509.Certificate{cert}
+	if isSelfSigned(cert) {
+		return chain, nil
+	}
+
+	// Look for the rest of intermediates skipping the root.
+	for {
+		// The Keychain stores the subject as an attribute, but it safes some of
+		// the values to uppercase. We cannot use the cert.RawIssuer to restrict
+		// more the search with KSecAttrSubjectKeyID and kSecAttrSubject. We
+		// would need to "normalize" it in the same way.
+		parent, err := loadCertificate("", nil, cert.AuthorityKeyId)
+		if err != nil || isSelfSigned(parent) || cert.CheckSignatureFrom(parent) != nil {
+			break
+		}
+		cert = parent
+		chain = append(chain, cert)
+	}
+
+	return chain, nil
+}
+
+// StoreCertificateChain stores a certificate chain in the Apple Keychain. There
+// is no need to provide a label in the URI as Apple will use the CommonName as
+// the default label, but if one is provided, the leaf certificate in the
+// Keychain will be updated with the given label:
+//
+// Valid names (URIs) are:
+//   - "" will use the common name as the label
+//   - "mackms:" will use the common
+//   - "mackms:label=my-label" will use "my-label"
+//   - "mackms:my-label" will use the "my-label"
+func (k *MacKMS) StoreCertificateChain(req *apiv1.StoreCertificateChainRequest) error {
+	// There's not really need to require the name as macOS will use the common
+	// name as default.
+	if len(req.CertificateChain) == 0 {
+		return fmt.Errorf("storeCertificateChainRequest 'certificateChain' cannot be empty")
+	}
+
+	// Do not require any parameter. Using mackms: is allowed as macOS will set
+	// the commonName as label.
+	u, err := parseCertURI(req.Name, false)
+	if err != nil {
+		return fmt.Errorf("mackms StoreCertificateChain failed: %w", err)
+	}
+
+	// Store the certificate and update the label if required
+	if err := storeCertificate(u.label, req.CertificateChain[0]); err != nil {
+		return fmt.Errorf("mackms StoreCertificateChain failed: %w", err)
+	}
+
+	// Store the rest of the chain but do not fail if already exists
+	for _, cert := range req.CertificateChain[1:] {
+		if err := storeCertificate("", cert); err != nil && !errors.Is(err, security.ErrAlreadyExists) {
+			return fmt.Errorf("mackms StoreCertificateChain failed: %w", err)
+		}
+	}
```

Comment:
Renewals work fine because the serial number is treated as part of the primary key, but both certificates will be present in the keychain. One of the things I mentioned in the PR that we can add is updating certificates, but it won't be part of this PR.